### PR TITLE
Fix links in providers documentation

### LINKED ISSUE #6964: Possible broken documentation in https://qiskit.org/documentation/apidoc/providers.html
### Information

- **Qiskit Terra version**:
- **Python version**:
- **Operating system**:

### What is the current behavior?
In the [providers documentation page](https://qiskit.org/documentation/apidoc/providers.html), some documentation redirects seem to be broken. 

- In the first paragraph _Providers Interface_, `Schedule` and `Backend` link to the respective classes, but `QuantumCircuit` is not linked.
- There are references to `BackendV2`, `BackendV34`, `AerBackend` classes, but they are not linked to any class documentation (for example in the paragraph _Version Policy Support_). Should these be linked to appropriate documentation pages?
-  In the paragraph _Writing a New Provider_, some references to `BackendV1` link to a documentation, but others don't.  Similarly, reference to `run()` is linked but `execute()` is not.
- In _Transpiler Interface_ (_Custom Basis Gates_),  step 2., reference to `HGate` documentation is broken. (In the line "...  and some commonly used single qubit rotation gates like the HGate and UGate that should be sufficient for the transpiler to translate any circuit into the custom basis gates.")
- There may be other such broken docs on this page. Perhaps this entire page should be revisited from the point of view of documentation linking.

### Steps to reproduce the problem
Hover with the mouse over the class names - at some places, they don't lead to the documentation.

### What is the expected behavior?
All references linked to docs.

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

This adds in some missing cross-references and fixes some typos in
existing crossrefs for the `qiskit.providers` documentation.  The class
`BackendV1` has a private method as part of its abstract API for
subclasses to implement, but `sphinx.ext.autosummary` won't summarise
private methods (without us having fairly annoying knock-on effects in
the rest of the documentation), so instead we just put in an explicit
`.. automethod` directive to get it.

### Details and comments

Fix #6964.

The `AerBackend` crossref won't show up in the Terra documentation alone, but (hopefully) should when it's compiled together with the Aer documentation in the metapackage release process.

### CODE REVIEW

File: qiskit/providers/__init__.py [hunk 1]
```diff
@@ -115,14 +116,14 @@
...
-:class:`~qiskit.provider.BackendV1` object provides both information describing
+method to get available :class:`~qiskit.providers.BackendV1` objects. The
+:class:`~qiskit.providers.BackendV1` object provides both information describing
 a backend and its operation for the :mod:`~qiskit.transpiler` so that circuits
 can be compiled to something that is optimized and can execute on the
 backend. It also provides the :meth:`~qiskit.providers.BackendV1.run` method which can
 run the :class:`~qiskit.circuit.QuantumCircuit` objects and/or
 :class:`~qiskit.pulse.Schedule` objects. This enables users and other Qiskit
-APIs, such as :func:`~qiskit.execute.execute` and higher level algorithms in
+APIs, such as :func:`~qiskit.execute_function.execute` and higher level algorithms in
```
Comment (1ucian0) [inline qiskit/providers/__init__.py:126]:
```suggestion
APIs, such as :func:`~qiskit.execute` and higher level algorithms in
```

**Review by 1ucian0 — COMMENTED** [review]
I think `execute` is meant to be used directly from `qiskit`, no from `qiskit.execute_function`. But I'm not fully sure.

File: qiskit/providers/__init__.py (hunk 1 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py:126]: What's in the PR is actually correct. It needs to be the full path for sphinx to resolve it correctly

Comment (jakelishman) [reply qiskit/providers/__init__.py:126]: This doesn't actually affect how it appears (the tilde `~` truncates it to the last word anyway), it's just for defining how Sphinx should find the symbol.  It looks things up in their canonical representation, so `qiskit.execute` doesn't resolve to a cross-ref-able symbol, unfortunately (at least in my local docs build).  `` :obj:`.execute` `` _might_ resolve correctly, but I wouldn't want to bet on it always resolving correctly with a generic name.

Sphinx works in very mysterious ways :(

**Comment by jakelishman** [comment]
@1ucian0: That's absolutely true - it is called like `qiskit.execute` - but Sphinx cross-refs are a bit mysterious.  The current form of the documentation renders like this, with the cross-reference correctly linked.

Screenshot 2021-08-31 at 16 14 25

File: qiskit/providers/__init__.py (hunk 1 above)
Comment (1ucian0) [reply qiskit/providers/__init__.py:126]: I was not aware of this. Great to learn it :)

**Review by 1ucian0 — APPROVED** [review]

**Pull request merged**